src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandIncludeProperty.cs

```diff
@@ -0,0 +1,215 @@
+
+        /// <summary>
+        /// the corresponding navigation source. maybe useless
+        /// </summary>
+        private IEdmNavigationSource _navigationSource;
+
+        /// <summary>
+        /// the path select item for this property.
+        /// for example: $select=abc or $select=NS.Type/abc
+        /// </summary>
+        private PathSelectItem _propertySelectItem;
+
+        /// <summary>
+        /// the sub $select and $expand for this property.
+        /// </summary>
+        private IList<SelectItem> _subSelectItems;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="SelectExpandIncludeProperty"/> class.
+        /// </summary>
+        /// <param name="propertySegment">The property segment that has this select expand item.</param>
+        public SelectExpandIncludeProperty(PropertySegment propertySegment)
+            : this(propertySegment, null)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="SelectExpandIncludeProperty"/> class.
+        /// </summary>
+        /// <param name="propertySegment">The property segment that has this select expand item.</param>
+        /// <param name="navigationSource">The targe navigation source of this property segment.</param>
+        public SelectExpandIncludeProperty(PropertySegment propertySegment, IEdmNavigationSource navigationSource)
+        {
+            if (propertySegment == null)
+            {
+                throw Error.ArgumentNull("propertySegment");
+            }
+
+            _propertySegment = propertySegment;
+            _navigationSource = navigationSource;
+        }
+
+        /// <summary>
+        /// Gets the merged path select item for this property, <see cref="PathSelectItem"/>.
+        /// </summary>
+        /// <returns>Null or the created <see cref="PathSelectItem"/>.</returns>
+        public PathSelectItem ToPathSelectItem()
+        {
+            if (_subSelectItems == null)
+            {
+                return _propertySelectItem;
+            }
+
+            // so, _subSelectItems is not null, merge the select and expand from the property into the _subSelectItems
+            bool isSelectAll = false;
+            if (_propertySelectItem != null && _propertySelectItem.SelectAndExpand != null)
+            {
+                // Retrieve the "IsSelectAll" from the property sub selectexpand clause.
+                isSelectAll = this._propertySelectItem.SelectAndExpand.AllSelected;
+                foreach (var selectItem in this._propertySelectItem.SelectAndExpand.SelectedItems)
+                {
+                    _subSelectItems.Add(selectItem);
+                }
+            }
+
+            if (isSelectAll)
+            {
+                // We do nothing here, because the property itself tells us to select all.
+                // Meanwhile, ODL doesn't allow $select=abc,abc(...), so it's safe to use "SelectAll".
+            }
+            else
```

Comment:
if (!isSelectAll)

Reply:
I leave the code like these for better understanding and maintenance. The C# compiler can optimize this if (..) clause (..)

---
In reply to: [352416627](https://github.com/OData/WebApi/pull/1941#discussion_r352416627) [](ancestors = 352416627)